exciting_environments/pmsm/pmsm_env.py

```diff
@@ -0,0 +1,922 @@
+                return d_y
+
+        term = diffrax.ODETerm(vector_field)
+        t0 = 0
+        t1 = action_stepsize * actions.shape[0]
+        y0 = tuple([i_d, i_q, eps])
+        saveat = diffrax.SaveAt(ts=jnp.linspace(t0, t1, 1 + int(t1 / obs_stepsize)))  #
+        y = diffrax.diffeqsolve(term, self._solver, t0, t1, dt0=obs_stepsize, y0=y0, args=args, saveat=saveat)
+
+        i_d_t = y.ys[0]
+        i_q_t = y.ys[1]
+        eps_t = y.ys[2]
+        obs_len = i_d_t.shape[0]
+
+        if properties.saturated:
+            torque_t = jax.vmap(self.currents_to_torque_saturated, in_axes=(0, 0, None))(i_d_t, i_q_t, properties)
+
+        else:
+            torque_t = jax.vmap(self.currents_to_torque, in_axes=(0, 0, None))(i_d_t, i_q_t, properties)
+
+        phys = self.PhysicalState(
+            u_d_buffer=jnp.zeros(obs_len),
+            u_q_buffer=jnp.zeros(obs_len),
+            epsilon=eps_t,
+            i_d=i_d_t,
+            i_q=i_q_t,
+            torque=torque_t,
+            omega_el=jnp.full(obs_len, init_state_phys.omega_el),
+        )
+        additions = None
+        ref = self.PhysicalState(
+            u_d_buffer=jnp.full(obs_len, jnp.nan),
+            u_q_buffer=jnp.full(obs_len, jnp.nan),
+            epsilon=jnp.full(obs_len, jnp.nan),
+            i_d=jnp.full(obs_len, jnp.nan),
+            i_q=jnp.full(obs_len, jnp.nan),
+            torque=jnp.full(obs_len, jnp.nan),
+            omega_el=jnp.full(obs_len, jnp.nan),
+        )
+        return self.State(
+            physical_state=phys, PRNGKey=jnp.full(obs_len, init_state.PRNGKey), additions=additions, reference=ref
+        )
+
+    def constraint_denormalization_ahead(self, actions, init_state, env_properties):
+        act_len = actions.shape[0]
+        phys = self.PhysicalState(
+            u_d_buffer=jnp.full(act_len, init_state.physical_state.u_d_buffer),
+            u_q_buffer=jnp.full(act_len, init_state.physical_state.u_q_buffer),
+            epsilon=jnp.full(act_len, init_state.physical_state.epsilon)
+            + jnp.linspace(0, self.tau * (act_len - 1), act_len) * init_state.physical_state.omega_el,
+            i_d=jnp.full(act_len, init_state.physical_state.i_d),
+            i_q=jnp.full(act_len, init_state.physical_state.i_q),
+            torque=jnp.full(act_len, init_state.physical_state.torque),
+            omega_el=jnp.full(act_len, init_state.physical_state.omega_el),
+        )
+        additions = None
+        ref = self.PhysicalState(
+            u_d_buffer=jnp.full(act_len, init_state.reference.u_d_buffer),
+            u_q_buffer=jnp.full(act_len, init_state.reference.u_q_buffer),
+            epsilon=jnp.full(act_len, init_state.reference.epsilon),
+            i_d=jnp.full(act_len, init_state.reference.i_d),
+            i_q=jnp.full(act_len, init_state.reference.i_q),
+            torque=jnp.full(act_len, init_state.reference.torque),
+            omega_el=jnp.full(act_len, init_state.reference.omega_el),
+        )
```

Comment:
Is there maybe a more efficient/cleaner/easier to understand way to do this using PyTree manipulation?